### ISSUE TITLE
Merging with `hadd` fails if there are more than `ulimit -n` files to merge

### ISSUE THREAD

**Issue by klannon** [matz-e/lobster#201]
Thanks to @annawoodard, it looks like the solution is to use `hadd -n 0`.  We need to test that this works as expected and them implement the fix.

**Comment by matz-e** [comment]
Keep in mind that `ulimit -n` may still limit the number of files to be merged.  Maybe we need #69 or a limitation on the number of files to be merged from the get-go?

**Comment by klannon** [comment]
Based on reading the source code for `hadd` (Hey, I didn't have time to work on it, but I can read things while I eat lunch!), when you set a file limit via the `-n N` option, `hadd` opens the files in batches of `N` until it's processed all the files.  So, rather than crashing, it should make it through any sized list.  Why this isn't the default behavior is beyond me.  Maybe there's something I'm missing, but it looks like it should work.

**Comment by matz-e** [comment]
Ah, dang!  You're right!  Helps to be able to read :facepalm:, and not just skim through the source of `hadd`.

**Comment by klannon** [comment]
So, I could not interactively replicate the crash.  I'm wondering now if it wasn't just bad luck landing on particular nodes where the # of files limit was particularly low.  (This seems to be a systems configurable parameter.)  In any case, I have verified that setting `-n N` does exactly what you'd expect:  `hadd` opens only that many files at a time.  In fact, it counts the output file as one open file (as I suspect it should), so it even accounts for that.  It should be a quick fix to add `-n 0` into the code, so I'll do that shortly.

**Comment by matz-e** [comment]
It may be that `hadd` fails for different reasons.  But I remember that I had problems merging large numbers of input files before - don't recall the failure mode anymore, though (may have been silent.)

**Comment by klannon** [comment]
Well, the error message in my log files seems pretty explicit:
```
hadd Source file 890: file:multilep_tree_6182.root
SysError in <TFile::TFile>: file multilep_tree_6182.root can not be opened for reading (Too many open files)
Error in <TFileMerger::AddFile>: cannot open file file:multilep_tree_6182.root
hadd exiting due to error in file:multilep_tree_6182.root
```
I suppose it's possible this was a red herring and the real problem was something else.  Whatever it was, the crash happened repeatedly on the same input.  In any case, using `-n 0` will make sure there is always a sensible limit on the number of simultaneously open files, which seems like a worthwhile precaution to take.

**Comment by annawoodard** [comment]
It's unfortunate that the issue isn't easily reproducible. At this point though I think there's nothing to do except try another run with `-n 0`. With heterogeneous resources, even that doesn't tell us anything for sure. That's part of the fun!

**Comment by klannon** [comment]
Well, with a little bit of effort I was at least able to extract the specific nodes that generated this error in the last run...
```
[klannon@earth test]$ zgrep -l 'SysError in <TFile::TFile>: file multilep_tree_6182.root can not be opened for reading (Too many open files)' /tmpscratch/users/klannon/ttH_13TeV_spring15_v3/*/failed/*/*/executable.log.gz | sed 's/executable.log.gz/job.log.gz/' | xargs zgrep =hostname= | awk '{print $NF}' | sort | uniq -c
     20 q16copt058.crc.nd.edu
     24 q16copt060.crc.nd.edu
      8 q16copt117.crc.nd.edu
```
It looks like it could be a feature of the `q16copt` cluster.  If we don't see a repeat, but we run large merge jobs on those nodes, we'll know something was fixed...

**Comment by annawoodard** [comment]
@klannon you remind me that it would be good to show this kind of information on the plotting page. It would be easy to implement some simple reporting about errorful hosts (more sophisticated reporting should probably wait for [ELK](https://www.elastic.co/products).) We've talked about this before, but it's never made it into an issue. I'll make one now.

**Comment by klannon** [comment]
Closed in commit ac1804c94c820c5c6758c441d956f21fec1c5f8, only I didn't know how to properly do that from a commit message until just now...
